### PR TITLE
Add OpenAPI specs

### PR DESCRIPTION
**Pull request by plars**
This uses schemas and apiflask to generate OpenAPI docs via a /docs link that you can browse to.  It covers most of the APIs along with the fields that they can accept and return. There are a few things that couldn't be fully specified because of the way the v1 api works.  For example, some things return string data, or return json but there are no predefined keys (queue/image mappings for example, use the queue name as the key). This should be considered in future revisions of the API to allow everything to be specified more easily.

I've handled all the changes needed for the unit tests to account for some data validation that happens automatically by using these schemas (such as 422 errors when we give it bad data). I've also tested this locally using a stress run that hits most of the api. However, we should definitely do a full integration test by deploying this to staging and put full runs through it before pushing to production.

### CODE REVIEW

File: src/api/schemas.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+"""
+
+from apiflask import Schema, fields
+from apiflask.validators import OneOf
+
+
+class AgentIn(Schema):
+    """Agent data input schema"""
+
+    state = fields.String(required=False)
```
[the remark this answers is not in the document]
Comment (plars) [reply src/api/schemas.py:27]: Yeah, I kinda liked being specific about it to make it really clear

File: src/api/schemas.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    parent_job_id = fields.String(required=False)
+    name = fields.String(required=False)
+    job_queue = fields.String(required=True)
+    global_timeout = fields.Integer(required=False)
+    output_timeout = fields.Integer(required=False)
+    allocation_timeout = fields.Integer(required=False)
+    provision_data = fields.Dict(required=False)
+    test_data = fields.Dict(required=False)
+    allocate_data = fields.Dict(required=False)
+    recover_data = fields.Dict(required=False)
```
[the remark this answers is not in the document]
Comment (plars) [reply src/api/schemas.py:53]: they are variable - some are dependent on the type of device. That's something to think about for later down the road with v2 api I think. For example, provisioning a device with muxpi you need to specify a url key. For maas, you need to specify a distro key, because it must already be imported and can't use a generic url.

File: src/api/v1.py [hunk 3]
```diff
@@ -55,24 +57,25 @@ def get_version():
...
         job_queue = ""
     if not job_queue:
         return "Invalid data or no job_queue specified\n", 400
 
     try:
-        job = job_builder(data)
+        job = job_builder(json_data)
     except ValueError:
-        return "Invalid job_id specified\n", 400
+        abort(400, message="Invalid job_id specified")
```
[the remark this answers is not in the document]
Comment (plars) [reply src/api/v1.py]: fixed, and also added a unit test to cover the corner case where the queue could be specified but is empty. I also changed this to use 422 instead of 400 since that's a little more specific.

File: src/api/v1.py [hunk 4]
```diff
@@ -106,6 +109,7 @@ def job_builder(data):
 
 
 @v1.get("/job")
+@v1.output(schemas.Job)
```
[the remark this answers is not in the document]
Comment (plars) [reply src/api/v1.py:112]: I do have a unit test that covers getting a job from an empty (or nonexistent) queue. It doesn't seem to care on output as much as input. Strange, yeah, I agree. Maybe it allows for some flexibility in responses for when there's no data?

[the remark this answers is not in the document]
Comment (omar-selo) [reply src/api/v1.py:112]: Looks like 

> [APIFlask will not validate the output data](https://apiflask.com/usage/#use-appoutput-to-format-response-data)

If so then this works, but I wonder if it outputs misleading openapi schema

**Review by omar-selo — APPROVED** [review]
Looks good. My only concern is whether open api spec for get /job endpoint shows that nothing will be returned if no job was found in the queue.

**Comment by plars** [comment]
It's true, It doesn't show that with 204 that it will return nothing (although it could be argued that 204, 404, etc should always return something that doesn't match the default output schema). I did try to see if I could give it multiple output decorators to convince it that 204 returned a different schema, but it only seems to take the first one

**Comment by omar-selo** [comment]
Looks like you can use `@app.doc` when you have multiple possible outputs. See Alert or Info box down in https://apiflask.com/usage/#use-appoutput-to-format-response-data

**Comment by plars** [comment]
Oh, excellent, that worked great! The docs for using @api.doc() are ironically kinda limited and don't have good examples, but I was able to sort out how to use it for this, and also how to use it for providing better docs and examples for some of the tricky sections that I mentioned before, such as when we return the advertised queues and images!
The other nice thing with this, is that for things like job where there's a schema for the valid response, I don't have to rewrite all of that as a json spec, I can just add on the empty version of it with 204 and it documents both in the output. Here's what it shows after adding this:
image

**Review by omar-selo — APPROVED** [review]
Nice looking great! Glad we could handle all those cases
